diffrax/_solver/foster_langevin_srk.py

```diff
@@ -0,0 +1,410 @@
+        (as opposed to via Taylor expansion). This function is then mapped over the
+        PyTree of gamma to compute the coefficients for the entire system.
+
+        **Arguments:**
+
+        - `h`: The time increment.
+        - `c`: A leaf of gamma.
+
+        **Returns:**
+
+        The SRK coefficients for the given leaf of gamma.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def _tay_coeffs_single(self, c: UnderdampedLangevinLeaf) -> _Coeffs:
+        r"""This method specifies how to compute the Taylor coefficients for a
+        single leaf of gamma. These coefficients are then used to compute the SRK
+        coefficients using the Taylor expansion. This function is then mapped over
+        the PyTree of gamma to compute the coefficients for the entire system.
+
+        **Arguments:**
+
+        - `c`: A leaf of gamma.
+
+        **Returns:**
+
+        The Taylor coefficients for the given leaf of gamma.
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    def _eval_taylor(h: RealScalarLike, tay_coeffs: _Coeffs) -> _Coeffs:
+        r"""Computes the SRK coefficients via the Taylor expansion, using the
+        precomputed Taylor coefficients and the step-size h."""
+        h = jnp.asarray(h, dtype=tay_coeffs.dtype)
+
+        def eval_taylor_fun(tay_leaf):
+            # jnp.polyval performs the Taylor expansion along
+            # the first axis, so we need to move the trailing
+            # axis to 0th position
+            transposed = jnp.moveaxis(tay_leaf, -1, 0)
+            assert transposed.shape[0] == 6
+            return jnp.polyval(transposed, h, unroll=6)
+
+        return jtu.tree_map(eval_taylor_fun, tay_coeffs)
+
+    def _recompute_coeffs(
+        self,
+        h: RealScalarLike,
+        gamma: UnderdampedLangevinX,
+        tay_coeffs: PyTree[_Coeffs],
+    ) -> _Coeffs:
+        r"""When h changes, the SRK coefficients (which depend on h) are recomputed
+        using this function."""
+        # Inner will record the tree structure of the coefficients
+        inner = sentinel = object()
+
+        def recompute_coeffs_leaf(c: UnderdampedLangevinLeaf, _tay_coeffs: _Coeffs):
+            # c is a leaf of gamma
+            # Depending on the size of h*gamma choose whether the Taylor expansion or
+            # direct computation is more accurate.
+            cond = h * c < self.taylor_threshold
+            tay_out = self._eval_taylor(h, _tay_coeffs)
+            if cond.ndim < jtu.tree_leaves(tay_out)[0].ndim:
+                # This happens when c is a scalar
+                cond = jnp.expand_dims(cond, axis=-1)
+
+            def select_tay_or_direct():
+                if jnp.ndim(c) == 0:
+                    direct_out = self._directly_compute_coeffs_leaf(h, c)
+                else:
+                    fun = lambda _c: self._directly_compute_coeffs_leaf(h, _c)
+                    direct_out = vmap(fun)(c)
```

Comment:
I'd add an `assert jnp.ndim(c) == 1` here, which I assume is what is going on.

Reply:
It's very good you wrote this here, made me realise this vmap is actually redundant. Originally LangevinSRK could only handle rank 0 and 1 tensors, so `self._directly_compute_coeffs_leaf` could only take a scalar `_c`. But now that it can handle arbitrary shapes, I can get rid of this.